README.md
# api-connector
Axios wrapper that simplifies JS API SDK development.  

[![npm version](https://badge.fury.io/js/api-connector.svg)](https://badge.fury.io/js/api-connector)
[![Build Status](https://travis-ci.com/iliyasappazov/api-connector.svg?branch=master)](https://travis-ci.com/iliyasappazov/api-connector)
[![Coverage Status](https://coveralls.io/repos/github/iliyasappazov/api-connector/badge.svg)](https://coveralls.io/github/iliyasappazov/api-connector)
[![dependencies Status](https://david-dm.org/iliyasappazov/api-connector/status.svg)](https://david-dm.org/iliyasappazov/api-connector)

## Getting started

### Installation
```npm install api-connector```

### Import
Import `api-connector`:
```js
import ApiConnector from 'api-connector';
// or
ApiConnector = require('api-connector').default;
```

### Example
Create `GET` request with async/await:
```js
const response = await ApiConnector.reqGet('https://example.com')
                                   .start();
```
Create `GET` request without a promise:
```js
ApiConnector.reqGet('https://example.com')
            .onOk(response => console.log(response))
            .startQuietly();
```
Combine both:
```js
const responseData = await ApiConnector.reqGet('https://example.com')
                                       .onOk(response => response.data)
                                       .start();
```

### ApiRequest modes: `simple`/`promised`
As you can see in the example above there are two ApiRequest modes:
+ **simple** - based purely on callbacks;
+ **promised** - returns a Promise that will be resolved with the result of the last `onOk` handler 
or rejected with the result of the last `onFail`/`onCancel`/`onError` handler.  
  
To start a **simple** request call the `startQuietly`/`startSingleQuietly` methods.

Example:
```js
const request = ApiConnector.reqGet('https://example.com')
                            .onOk(response => console.log(response));

// start a promised ApiRequest:
await request.start()

// start a simple ApiRequest:
request.startQuietly();

// start a simple ApiRequest and then convert it to promised:
request.startQuietly();
await request.genPromise();
```

### Axios configuration
ApiConnector simply wraps Axios requests thus all Axios configurations are still available.   
You can create an ApiConnector instance with common axios configurations:
```js
const generalAxiosConfigs = { base_url: 'http://localhost' };
const connector = ApiConnector.create(generalAxiosConfigs);
```
And/or pass the Axios configuration object directly to the request through `config` parameter:
```js
const axiosConfigs = { timeout: 1000 };
const response = await connector.reqGet('/test', {param1: 'test'}, axiosConfigs).start();
```
Or pass configured Axios instance directly to the request through `apiConfigs` parameter:
```js
const response = await connector.reqGet('/test', {param1: 'test'}, {}, { axios: anyAxiosInstance }).start();
```

### Response validation
ApiConnector has a mechanism to determine the successfulness of the response.  
By default all completed responses are considered as successful.  
You can change this logic by setting a *validation function*.
This function should accept the Axios response object and return `true` if response is successful and `false` otherwise.
Specify *validation function* for ApiConnector instance:
```js
const testConnector = ApiConnector.create(
  {baseURL: 'http://localhost/api'},
  response => response.data.result === 'OKAY'
);
const response = await testConnector.reqGet('/test')
                    .onOk(() => console.log('Success')) // will be called on { result: 'OKAY' ... } response
                    .onFail(() => console.log('Fail'))
                    .start();
```
or specify *validation function* only for a single request:
```js
const axiosInstance = Axios.create({ baseURL: 'http://localhost/api/v2'});
const response = await testConnector.reqGet(
                        '/test', 
                        params, 
                        null,
                        { 
                          axios: axiosInstance,
                          validateFunc: response => response.data.result === 'OK' 
                        }
                    )
                    .onOk(() => console.log('Success')) // will be called on { result: 'OK' ... } response
                    .onFail(() => console.log('Fail'))
                    .start();
```

### Event handling
#### Basic events
ApiConnector allows you to attach callbacks to the following **basic** events:
+ `onOk` - *successful* API response;
+ `onFail` - *failed* API response;
+ `onCancel` - request cancellation;
+ `onError` - exception;
+ `onStatus` - status code.

There are two ways to attach callbacks to any of the above events. The simplest one - call a method with the same name:
```js
ApiConnector.reqGet('http://localhost')
            .onOk(processResponse)
            .onFail(alertError)
            .onCancel(alertError)
            .onError(alertError)
            .onStatus(logout, 401, 403)
```
Another way to attach callbacks - use `onAny` method:
```js
ApiConnector.reqGet('http://localhost')
            .onAny(processResponse, 'onOk')
            .onAny(alertError, 'onFail', 'onCancel', 'onError')
            .onAny(authorizeResponse, 'onStatus=[401, 403]');
```

#### Combined events
For convenience there are **combined** events that represents a different combinations of the **basic** events:
+ `onResponse` - combines `onOk` and `onFail` events
+ `onAnyError` - combines `onFail`, `onCancel`, `onError` events
+ `then` - combines `onOk`, `onFail`, `onCancel`, `onError` events 

Callbacks to the **combined** events could be attached only by calling a method with the same name: 
```js
ApiConnector.reqGet('http://localhost')
            .onResponse(processResponse)
            .onAnyError(alertError)
            .then(sayGoodbye);
```
the code above is equivalent to:
```js
ApiConnector.reqGet('http://localhost')
            .onAny(processResponse, 'onOk', 'onFail')
            .onAny(alertError, 'onFail', 'onCancel', 'onError')
            .onAny(sayGoodbye, 'onOk', 'onFail', 'onCancel', 'onError');
```

#### Multiple handlers
You could attach multiple callbacks on the same event, all these callbacks will be fired in order in which they were added.  
The result from the previous callback will be passed to the next one:
```js
const result = await ApiConnector.reqGet('http://localhost')
                                 .onOk(() => 'Hello,')
                                 .onOk(hello => hello + ' World')
                                 .then(helloWorld => helloWorld + '!')
                                 .start();
console.log(result); // 'Hello, world!'                                 
```

#### Post-handlers
Event callbacks could be attached after a request start:
```js
const request = ApiConnector.reqGet('http://localhost');
await request.onOk(() => 'Hello, World!')
             .start();
request.onOk(res => console.log(res));  // 'Hello, World!'
```

#### Promised ApiRequest
`Promised` ApiRequest returns a Promise that:
+ resolves with the result of the last `onOk` event handler;
+ on `onFail` event throws the following Error: 
    ```js
    {
      isFail: true,
      data: any // result of the last onFail callback 
    }
    ```
+ on `onCancel` event throws the following Error:
     ```js
     {
       isCancel: true,
       data: any // result of the last onCancel callback 
     }
     ```
+ on `onError` event throws the result of the last `onError` callback.

Example:
```js
try {
  const response = await ApiConnector.reqGet('http://localhost')
                                     .onOk(() => 'Ok')
                                     .onFail(() => 'Failed')
                                     .onCancel(() => 'Cancelled')
                                     .onError(() => 'Error')
                                     .start();
  console.log(response); // 'Ok'
} catch (err) {
  if (err.isFail) {
    console.log(err.data); // 'Failed'
  } else if (err.isCancel) {
    console.log(err.data); // 'Cancelled'
  } else {
    console.log(err);  // 'Error'
  }
}
```

### Request cancellation
#### Manual request cancellation
You could cancel request manually by calling the `cancel` method:
```js
const request = ApiConnector.reqGet('http://localhost')
                            .onCancel(() => console.log('Cancelled'));
setTimeout(() => request.cancel(), 100);
request.start();
```
`startQuietly`/`startSingleQuietly` methods returns Axios request cancellation functions:
```js
const cancel = ApiConnector.reqGet('http://localhost')
                           .onCancel(() => console.log('Cancelled'))
                           .startQuietly();
cancel();
```

#### Automatic request cancellation
`startSingle` method performs new Axios request and cancels previous pending requests with the same method, url and identifier:
```js
ApiConnector.reqGet('http://localhost', { param1: 1}).start('same');
ApiConnector.reqGet('http://localhost', { param1: 1}).startSingle('same');  // will cancel previous request
ApiConnector.reqGet('http://localhost', { param1: 1}).startSingle('different');  // will NOT cancel any requests, because request identifiers are different
ApiConnector.reqGet('http://localhost', { param1: 2}).startSingle('different');  // will NOT cancel any requests, because request parameters are different
```

### SDK example
Assume:
1. you have an API with base url: `https://example.com/api`;
2. all successful responses have the following format: `{ status: 'success', result: *}`;
3. all failed responses have the following format: `{ status: 'fail', message: ''}`;
4. there are GET `/users` and DELETE `/user?name={name}` endpoints;
5. SDK users should get only `result` response field value;
6. on error SDK should log error or `message` response field.

Sample SDK:
```js
import ApiConnector from 'api-connector';

const api = ApiConnector.create(
        { baseURL: 'https://example.com/api' }, // 1.
        response => response.data.status === 'success'  // 2.
    );

function addCommonHandlers(apiRequest) {
    return apiRequest.onOk(response => response.data.result)  // 5.
                     .onFail(response => response.data.message);  // 6.
}

const SDK = {
    getUsers: async () => await addCommonHandlers(api.reqGet('/users')).start(),
    deleteUser: async (username) => await addCommonHandlers(api.reqDelete('/user', { name: username })),
}; // 4.

try {
    const users = await SDK.getUsers();
    console.log('Total users count: ', users.length);

    await SDK.deleteUser('John');
} catch (e) {
    if (e.isFail) {
      console.log(e.data);
    } else if (!e.isCancel) {
      console.log(e);
    }
}
```


### ApiConnector API

#### .create(configs?, validateFunc?)
Create ApiConnector with Axios configs:
```js
myApiConn = ApiConn.create({baseURL: 'https://example.com'});
```
Provided validate function will be used in response status evaluation:
```js
myApiConn = ApiConn.create({}, response => response.data.status === 'ok');
```

#### .request(axiosConfig, apiConfig)
Creates and returns `ApiRequest` using *Axios request* configs and *ApiConnector request* configs.
```js
req = ApiConn.request({method: 'GET', url: 'https://example.com'}, { validateFunc: response => !!response.data });
```

#### .req* Aliases
For convenience aliases have been provided for some request methods:

##### .reqGet(url, params?, axiosConfig?, apiConfig?)
##### .reqPost(url, data?, params?, axiosConfig?, apiConfig?)
##### .reqPatch(url, data?, params?, axiosConfig?, apiConfig?)
##### .reqPut(url, data?, params?, axiosConfig?, apiConfig?)
##### .reqDelete(url, params?, axiosConfig?, apiConfig?)

These methods just calls `.request` method, e.g, `.reqPost` is equal to `ApiConn.request({method: 'POST', url, data, params, ...config}, apiConfig)`. 


### ApiRequest API

#### start(identifier='')
Performs http request.
Returns a Promise:
+ `onOk` event will resolve the Promise with a data returned by the last `onOk` handler.
+ `onFail` event will reject the Promise with the following Error: 
    ```js
    {
      isFail: true,
      data: any // result of the last onFail callback 
    }
    ```
+ `onCancel` event will reject the Promise with the following Error:
     ```js
     {
       isCancel: true,
       data: any // result of the last onCancel callback 
     }
     ```
+ `onError` event will reject the Promise with the result of the last `onError` callback.

#### startSingle(identifier='')
Similar to `start` method but cancels previous pending requests with the same *method*, *url* and *identifier*.

#### .on* event callbacks
You can attach callbacks to the following events
+ *successful* API response (`.onOk`)
+ *failed* API response (`.onFail`)
+ any API response (`.onResponse`)
+ request cancellation (`.onCancel`)
+ exception (`.onError`)
+ after any of the above (`.then`)
+ after *failed* API response, or cancellation, or exception (`onAnyError`)
+ status code (`.onStatus`)
+ any event combination (`.onAny`)

##### .onAny(callback, ...eventNames)
Attaches callback to all given events.  
Supported event names: `'onOk'`, `'onFail'`, `'onCancel'`, `'onError'`, `'onStatus={status}'`, `'onStatus=[{status1}, {status2}]'`.  
All unsupported event names will be ignored.

Attach callback to request cancellation and response with status code 401:
```js
await ApiConn.reqGet('https://example.com')
             .onAny(callback, 'onCancel', 'onStatus=401')
             .start();
```

##### .onStatus(callback, ...statuses)
Adds callback to response with exact status code.

Attach callback to responses with 201 and 202 status codes.
```js
await ApiConn.reqGet('https://example.com')
             .onStatus(callback, 201, '202')
             .start();
```

## License
[MIT License](https://github.com/iliyasappazov/api-connector/blob/master/LICENSE)


CHANGELOG.md
# 1.2.1
+ **fix:** update dependencies.

# 1.2.0

+ **feature:** remove `promise` parameter from `start` and `startSingle` methods.
+ **feature:** add `startQuietly` and `startSingleQuietly` methods.
+ **fix:** update dependencies.

# 1.1.0

+ **feature:** add ability to specify *validation function* and *Axios instance* for a single request.

# 1.0.0

+ Initial Release


test/_common.js
/**
 * Calls `request.start()` and returns Promise that will be resolved after request completed.
 *
 * @param {ApiRequest} request
 * @return {Promise} promise that will be returned on
 */
export function testRequest(request) {
  return new Promise((resolve) => {
    request.then(resolve).startQuietly();
  });
}

/**
 * Calls `request.startSingle()` and returns Promise that will be resolved after request completed.
 *
 * @param {ApiRequest} request
 * @return {Promise} promise that will be returned on
 */
export function testRequestSingle(request) {
  return new Promise((resolve) => {
    request.then(resolve).startSingleQuietly();
  });
}


test/_server.js
import http from 'http';
import url from 'url';
import net from 'net';

export function getFreePort() {
  return new Promise((resolve) => {
    const server = net.createServer();
    server.listen(() => {
      const { port } = server.address();
      server.close(() => resolve(port));
    });
  });
}

/**
 * Creates an http server.
 * Server simply returns request info in the following format: `{url, method, params, data}`.
 * If `status` url parameter is given - its value will be used as response status code.
 * If `timeout` url parameter is given - its value will be used to determine response time
 *
 * @param {number} port
 * @return {http.Server}
 */
export function createServer(port) {
  const server = http.createServer((req, res) => {
    const response = {
      url: req.url,
      method: req.method,
      params: url.parse(req.url, true).query,
    };
    let body = [];
    req.on('data', (chunk) => {
      body.push(chunk);
    });

    req.on('end', () => {
      if (body.length > 0) {
        body = Buffer.concat(body).toString();
        try {
          response.data = JSON.parse(body);
        } catch (e) {
          response.data = body;
        }
      }
      setTimeout(() => {
        res.writeHead(response.params.status || 200);
        res.write(JSON.stringify(response));
        res.end();
      }, response.params.timeout || 0);
    });
  });
  server.listen(port, 'localhost');
  return server;
}


test/pipe.test.js
import test from 'ava';
import { Pipe } from '../lib/utils';

test('piped callbacks are called sequentially', (t) => {
  const initialResponse = {
    data: {
      inner: 'value',
    },
  };

  const pipe = new Pipe();
  pipe.join((response) => response.data)
    .join((data) => data.inner)
    .join((inner) => t.is(inner, 'value'));

  pipe.process(initialResponse);
});

test(
  'after a repeated call, the pipe process method continues processing, starting with the newly added callbacks',
  (t) => {
    const data = { value: 1 };
    let innerData = data;
    for (let i = 0; i < 5; i += 1) {
      innerData.inner = { value: innerData.value + 1 };
      innerData = innerData.inner;
    }

    const handler = (curData) => curData.inner;

    const pipe = new Pipe();
    pipe.join(handler)
      .join(handler)
      .join(handler);

    pipe.process(data);

    pipe.join((curData) => {
      t.is(curData.value, 4);
      return curData.inner;
    });
    pipe.join((curData) => {
      t.is(curData.value, 5);
    });
    pipe.process();
  },
);


lib/utils.js
/**
 * Pipe allows to build conveyor of handler functions to process any data.
 * Pipe passes data through all handler functions in such a way that each handler takes a data from the previous handler's output.
 */
export class Pipe {
  constructor() {
    this.pipe = [];
    this.curData = null;
    this.curStep = 0;
  }

  /**
   * Attaches a handler function to the pipe.
   *
   * @param {function} handler
   * @return {Pipe}
   */
  join(handler) {
    this.pipe.push(handler);
    return this;
  }

  /**
   * Sequentially calls handler functions from the pipe.
   * The first handler function is called with the given data object.
   * Each subsequent handler takes a data from the previous handler's output.
   * If pipe has already been processed:
   * + calls only newly connected handlers
   * + ignores initialData parameter
   *
   * @param {*} initialData - data to be passed to the first handler.
   */
  process(initialData) {
    if (this.curStep === 0) {
      this.curData = initialData;
    }
    for (; this.curStep < this.pipe.length; this.curStep += 1) {
      this.curData = this.pipe[this.curStep](this.curData);
    }
  }
}

export function hashCode(str) {
  let hash = 0;
  if (str.length === 0) return hash;
  for (let i = 0; i < str.length; i += 1) {
    /* eslint-disable no-bitwise */
    const chr = str.charCodeAt(i);
    hash = ((hash << 5) - hash) + chr;
    hash |= 0; // Convert to 32bit integer
    /* eslint-enable no-bitwise */
  }
  return hash;
}


lib/ApiConnector.js
import Axios from 'axios';
import ApiRequest from './ApiRequest';

export default class ApiConnector {
  /**
   * Creates api connector instance
   * @param {AxiosRequestConfig} axiosConfigs - Axios configs
   * @param {Function} validateFunc - function that decides whether response is Ok or Failed,
   *                                  accepts Axios response object, should return true if response is Ok
   */
  constructor(axiosConfigs = {}, validateFunc = null) {
    this.validateFunc = validateFunc;
    this.axios = Axios.create(axiosConfigs);
    /**
     * Dictionary with pending request cancel functions.
     * Keys are hashes of request method, url and id.
     * Values are another dictionaries - with request ts as keys and request cancel functions as values.
     *
     * @example
     * {
     *      someHashString: {
     *          1520000000: cancelFunction
     *          1520000001: cancelFunction
     *      }
     * }
     */
    this.pendingRequests = {};
  }

  /**
   * Creates GET request.
   *
   * @param {string} url - API endpoint url
   * @param {{}} params - request uri parameters
   * @param {AxiosRequestConfig} axiosConfigs - Axios request configuration
   * @param {ApiRequestConfig} apiConfigs - ApiRequest configuration
   * @returns {ApiRequest} request promise
   */
  reqGet(url, params = {}, axiosConfigs = {}, apiConfigs = {}) {
    return this.request({
      method: 'GET', url, params, ...axiosConfigs,
    }, apiConfigs);
  }

  /**
   * Creates POST request.
   *
   * @param {string} url - API endpoint url
   * @param {{}} data - request body data
   * @param {{}} params - request uri parameters
   * @param {AxiosRequestConfig} axiosConfigs - Axios request configuration
   * @param {ApiRequestConfig} apiConfigs - ApiRequest configuration
   * @returns {ApiRequest} request promise
   */
  reqPost(url, data = undefined, params = {}, axiosConfigs = {}, apiConfigs = {}) {
    return this.request({
      method: 'POST', url, data, params, ...axiosConfigs,
    }, apiConfigs);
  }

  /**
   * Creates PATCH request.
   *
   * @param {string} url - API endpoint url
   * @param {{}} data - request body data
   * @param {{}} params - request uri parameters
   * @param {AxiosRequestConfig} axiosConfigs - Axios request configuration
   * @param {ApiRequestConfig} apiConfigs - ApiRequest configuration
   * @returns {ApiRequest} request promise
   */
  reqPatch(url, data = undefined, params = {}, axiosConfigs = {}, apiConfigs = {}) {
    return this.request({
      method: 'PATCH', url, data, params, ...axiosConfigs,
    }, apiConfigs);
  }

  /**
   * Creates PUT request.
   *
   * @param {string} url - API endpoint url
   * @param {{}} data - request body data
   * @param {{}} params - request uri parameters
   * @param {AxiosRequestConfig} axiosConfigs - Axios request configuration
   * @param {ApiRequestConfig} apiConfigs - ApiRequest configuration
   * @returns {ApiRequest} request promise
   */
  reqPut(url, data = undefined, params = {}, axiosConfigs = {}, apiConfigs = {}) {
    return this.request({
      method: 'PUT', url, data, params, ...axiosConfigs,
    }, apiConfigs);
  }

  /**
   * Creates DELETE request.
   *
   * @param {string} url - API endpoint url
   * @param {{}} params - request uri parameters
   * @param {AxiosRequestConfig} axiosConfigs - Axios request configuration
   * @param {ApiRequestConfig} apiConfigs - ApiRequest configuration
   * @returns {ApiRequest} request promise
   */
  reqDelete(url, params = {}, axiosConfigs = {}, apiConfigs = {}) {
    return this.request({
      method: 'DELETE', url, params, ...axiosConfigs,
    }, apiConfigs);
  }

  /**
   * Creates ApiRequest
   *
   * @param {AxiosRequestConfig} axiosConfigs - Axios configs
   * @param {ApiRequestConfig} apiConfigs - ApiRequest configuration
   * @returns {ApiRequest} request promise
   */
  request(axiosConfigs, apiConfigs) {
    const curConfigs = {
      axios: this.axios,
      validateFunc: this.validateFunc,
      ...apiConfigs,
    };
    return new ApiRequest(curConfigs.axios, this.pendingRequests, curConfigs.validateFunc, axiosConfigs);
  }
}


index.d.ts
import {AxiosResponse, Canceler, AxiosRequestConfig, AxiosInstance} from "axios";

/**
 * Function that decides whether response is *successful* or *failed*,
 * accepts Axios response object, should return `true` if response is *successful*.
 */
export type ValidateResponseFunc<ResponseData> = (data: AxiosResponse<ResponseData>) => boolean;

/**
 * Dictionary with pending request cancel functions.
 * Keys are hashes of request method, url and id.
 * Values are dictionaries with request ts as keys and request cancel functions as values.
 *
 * @example
 * {
 *      someHashString: {
 *          1520000000: cancelFunction
 *          1520000001: cancelFunction
 *      }
 * }
 */
export type PendingRequestsDict = {
    [requestHash: string]: { [requestId: string]: Canceler };
}

export type EventCallbackFunc<DataIn, DataOut = DataIn> = (data: DataIn) => DataOut;

export type RequestEventNames = 'onOk' | 'onFail' | 'onCancel' | 'onError' | `onStatus=${number}`;

export type ApiRequestError<Fail, Cancel> = { isCancel: true, data: Cancel }
    | { isFail: true, data: Fail }
    | { data: any };

/**
 * ApiRequest creates and performs cancellable http request.
 * ApiRequest is able to attach callbacks to the following events:
 * + *successful* API response (`.onOk`)
 * + *failed* API response (`.onFail`)
 * + any API response (`.onResponse`)
 * + request cancellation (`.onCancel`)
 * + exception (`.onError`)
 * + after any of the above (`.then`)
 * + after *failed* API response, or cancellation, or exception (`onAnyError`)
 * + status code (`.onStatus`)
 * + any event combination (`.onAny`)
 */
export interface ApiRequest<Ok, Fail=Ok, Cancel = any> {
    /**
     * Stores configurations and initializes callback pipes.
     * Validate function will be used to evaluate whether response is successful or not.
     * Axios response object will be passed to the function,
     * and if `true` is returned the response is said to be *successful*.
     * If the function is `null` any response will be considered as *successful*.
     *
     * @param axios - Axios instance
     * @param pendingRequests  - dictionary with pending request cancel functions
     * @param validateFunc - function that decides whether a response is *successful* or *failed*
     * @param configs  - Axios configs
     */
    <In>(
        axios: AxiosInstance,
        pendingRequests: PendingRequestsDict,
        validateFunc: ValidateResponseFunc<In>,
        configs: AxiosRequestConfig
    ): ApiRequest<In>;

    /**
     * Attaches the callback to all specified events.
     * Supported event names: `"onOk"`, `"onFail"`, `"onCancel"`, `"onError"`, `"onStatus={status}"`.
     * All unsupported event names will be ignored.
     * Returns the same request.
     */
    onAny<OkOut = Ok, FailOut = Fail, CancelOut = Cancel>(
        callback: EventCallbackFunc<Ok | Fail | Cancel | any, OkOut | FailOut | CancelOut>,
        ...eventNames: RequestEventNames[]
    ): ApiRequest<OkOut, FailOut, CancelOut>;

    /**
     * Adds callback to *successful* response.
     * Returns the same request.
     */
    onOk<OkOut = Ok>(callback: EventCallbackFunc<Ok, OkOut>)
        : ApiRequest<OkOut, Fail, Cancel>;

    /**
     * Adds callback to *failed* response.
     * Returns the same request.
     */
    onFail<FailOut = Fail>(callback: EventCallbackFunc<Fail, FailOut>)
        : ApiRequest<Ok, FailOut, Cancel>;

    /**
     * Adds callback to any response.
     * Returns the same request.
     */
    onResponse<Out = Ok | Fail>(callback: EventCallbackFunc<Ok | Fail, Out>)
        : ApiRequest<Out, Out, Cancel>;

    /**
     * Adds callback to request cancellation.
     * Returns the same request.
     */
    onCancel<CancelOut = Cancel>(callback: EventCallbackFunc<Cancel, CancelOut>)
        : ApiRequest<Ok, Fail, CancelOut>;

    /**
     * Adds callback function that will be called on exception rise.
     * Returns the same request.
     */
    onError(callback: EventCallbackFunc<any, any>): ApiRequest<Ok, Fail, Cancel>;

    /**
     * Adds callback that will be fired last on any request result.
     * Returns the same request.
     */
    then<Out = Ok | Fail | Cancel>(callback: EventCallbackFunc<Ok | Fail | Cancel, Out>)
        : ApiRequest<Out, Out, Out>;

    /**
     * Adds callback to response with exact status code.
     * Returns the same request.
     */
    onStatus(callback: EventCallbackFunc<Ok | Fail | Cancel, any>, ...statuses: number[])
        : ApiRequest<Ok, Fail, Cancel>;

    /**
     * Adds callback that will be fired on any not Ok results:
     * + fail
     * + error
     * + cancel
     *
     * Returns the same request.
     */
    onAnyError<Out = Fail | Cancel>(callback: EventCallbackFunc<Fail | Cancel | any, Out>)
        : ApiRequest<Ok, Out, Out>;

    /**
     * Cancels requests if it is in pending state.
     */
    cancel(): void;

    /**
     * Creates and performs http request.
     * Cancels previous pending requests with the same method, url and identifier.
     *
     * Possible promise outcomes:
     * + `onOk` event will resolve the Promise with a data returned by the last `onOk` handler.
     * + `onFail` event will reject the Promise with an error with `isFail` property set `true`
     * and `data` property with a data returned by the last `onFail` handler.
     * + `onCancel` event will reject the Promise with an error with `isCancel` property set `true`
     * and `data` property with a data returned by the last `onCancel` handler.
     * + `onError` error will reject the Promise with a data returned by the last `onError` handler.
     *
     * @param identifier - some identifier
     */
    startSingle(identifier?: string): Promise<Ok>;

    /**
     * Performs http request.
     *
     * Possible promise outcomes:
     * + `onOk` event will resolve the Promise with a data returned by the last `onOk` handler.
     * + `onFail` event will reject the Promise with an error with `isFail` property set `true`
     * and `data` property with a data returned by the last `onFail` handler.
     * + `onCancel` event will reject the Promise with an error with `isCancel` property set `true`
     * and `data` property with a data returned by the last `onCancel` handler.
     * + `onError` error will reject the Promise with a data returned by the last `onError` handler.
     *
     * @param identifier - some identifier
     */
    start(identifier?: string): Promise<Ok>;

    /**
     * Creates and performs http request.
     * Cancels previous pending requests with the same method, url and identifier.
     * Returns Canceler function.
     *
     * @param {string|number} identifier - some identifier
     */
    startSingleQuietly(identifier?: string): Canceler

    /**
     * Performs http request.
     * Returns Canceler function.
     *
     * @param {string|number} identifier - some identifier
     */
    startQuietly(identifier?: string): Canceler;
}

export type ApiRequestConfig<ResponseData> = {
    axios: AxiosInstance;
    validateFunc: ValidateResponseFunc<ResponseData>;
}

export interface ApiConnector<ApiResponseData = any> {
    (axiosConfigs?: AxiosRequestConfig, validateFunc?: ValidateResponseFunc<ApiResponseData>)
        : ApiConnector<ApiResponseData>;

    validateFunc: ValidateResponseFunc<ApiResponseData>;
    axios: AxiosInstance;

    /**
     * Creates GET request.
     *
     * @param url - API endpoint url
     * @param params - request uri parameters
     * @param axiosConfigs - Axios request configuration
     * @param apiConfigs - Axios request configuration
     */
    reqGet<ResponseData = ApiResponseData>(
        url: string,
        params?: any,
        axiosConfigs?: AxiosRequestConfig,
        apiConfigs?: ApiRequestConfig<ResponseData>
    ): ApiRequest<AxiosResponse<ResponseData>>;

    /**
     * Creates POST request.
     *
     * @param url - API endpoint url
     * @param data - request body data
     * @param params - request uri parameters
     * @param axiosConfigs - Axios request configuration
     * @param apiConfigs - Axios request configuration
     */
    reqPost<ResponseData = ApiResponseData>(
        url: string,
        data?: any,
        params?: any,
        axiosConfigs?: AxiosRequestConfig,
        apiConfigs?: ApiRequestConfig<ResponseData>
    ): ApiRequest<AxiosResponse<ResponseData>>;

    /**
     * Creates PATCH request.
     *
     * @param url - API endpoint url
     * @param data - request body data
     * @param params - request uri parameters
     * @param axiosConfigs - Axios request configuration
     * @param apiConfigs - Axios request configuration
     */
    reqPatch<ResponseData = ApiResponseData>(
        url: string,
        data?: any,
        params?: any,
        axiosConfigs?: AxiosRequestConfig,
        apiConfigs?: ApiRequestConfig<ResponseData>
    ): ApiRequest<AxiosResponse<ResponseData>>;

    /**
     * Creates PUT request.
     *
     * @param url - API endpoint url
     * @param data - request body data
     * @param params - request uri parameters
     * @param axiosConfigs - Axios request configuration
     * @param apiConfigs - Axios request configuration
     */
    reqPut<ResponseData = ApiResponseData>(
        url: string,
        data?: any,
        params?: any,
        axiosConfigs?: AxiosRequestConfig,
        apiConfigs?: ApiRequestConfig<ResponseData>
    ): ApiRequest<AxiosResponse<ResponseData>>;

    /**
     * Creates DELETE request.
     *
     * @param url - API endpoint url
     * @param params - request uri parameters
     * @param axiosConfigs - Axios request configuration
     * @param apiConfigs - Axios request configuration
     */
    reqDelete<ResponseData = ApiResponseData>(
        url: string,
        params?: any,
        axiosConfigs?: AxiosRequestConfig,
        apiConfigs?: ApiRequestConfig<ResponseData>
    ): ApiRequest<AxiosResponse<ResponseData>>;

    /**
     * Creates ApiRequest
     *
     * @param axiosConfigs - Axios request configuration
     * @param apiConfigs - Axios request configuration
     */
    request<ResponseData = ApiResponseData>(
        axiosConfigs?: AxiosRequestConfig,
        apiConfigs?: ApiRequestConfig<ResponseData>
    ): ApiRequest<AxiosResponse<ResponseData>>;
}

export interface ApiConnectorStatic extends ApiConnector {
    create<ResponseData = any>(
        axiosConfigs?: AxiosRequestConfig,
        validateFunc?: ValidateResponseFunc<ResponseData>
    ): ApiConnector<ResponseData>;
}

declare const Default: ApiConnectorStatic;

export default Default;
